### PR TITLE
APP-1631 : Scenarios to Approve and Deny attachment

### PR DESCRIPTION
**Pull request by crepache**

### CODE REVIEW

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    assertEquals(state.toUpperCase(), getValueFromParam(message.get().getEntityData(), STATE));
+  }
+
+  private String getValueFromParam(String entityDate, String key) throws HealthCheckFailedException {
+    JsonNode node = null;
+
+    try {
+      node = MAPPER.readTree(entityDate);
+    } catch (IOException e) {
+      throw new HealthCheckFailedException("Failed to read response entity.");
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Why HealthCheckFailedException?

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    return node.get("makerchecker").get(key).asText();
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: It should be a constant field.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  @Then("$agent can verify the attachment is $state")
+  public void verifyStateOfMakerChecker(String agent, String state)
+      throws MessagesException, HealthCheckFailedException {
+    SymUser userAgent = userHelper.getUser(agent.toUpperCase());
+    Optional<SymMessage> message = messageHelper.getLatestTicketMessage(userAgent, initialTime);
+
+    assertEquals(state.toUpperCase(), getValueFromParam(message.get().getEntityData(), STATE));
+  }
+
+  private String getValueFromParam(String entityDate, String key) throws HealthCheckFailedException {
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Why entityDate? The method might be renamed to 'getValueFromEntityData'

Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Javadoc

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  @When("$agent agent approve the attachment")
+  public void approveAttachment(String agent)
+      throws MessagesException, HealthCheckFailedException, MalformedURLException {
+    SymUser userAgent = userHelper.getUser(agent.toUpperCase());
+    Optional<SymMessage> message = messageHelper.getLatestTicketMessage(userAgent, initialTime);
+
+    if (!message.isPresent()) {
+      throw new TicketRoomNotFoundException("Message not found");
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Why TicketRoomNotFoundException?

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  @When("$agent agent deny the attachment")
+  public void denyAttachment(String agent)
+      throws MessagesException, HealthCheckFailedException, MalformedURLException {
+    SymUser userAgent = userHelper.getUser(agent.toUpperCase());
+    Optional<SymMessage> message = messageHelper.getLatestTicketMessage(userAgent, initialTime);
+
+    if (!message.isPresent()) {
+      throw new TicketRoomNotFoundException("Message not found");
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Why TicketRoomNotFoundException?

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+  private static final String ATTACHMENTS_DIR = "attachment/";
+
+  private static final String APPROVE_URL = "approveUrl";
+
+  private static final String DENY_URL = "denyUrl";
+
+  private static final String STATE = "state";
+
+  private Long initialTime = 0L;
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Probably you should update this variable somewhere.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/listener/PrepareEnvironmentListener.java [hunk 7]
```diff
@@ -59,6 +64,11 @@ public void beforeTestClass(TestContext testContext) throws Exception {
     testContext.markApplicationContextDirty(DirtiesContext.HierarchyMode.CURRENT_LEVEL);
   }
 
+  private void setGroupId() {
+    String groupId = StringUtils.abbreviate(UUID.randomUUID().toString(), 20);
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/listener/PrepareEnvironmentListener.java]: 20 chars? You might reduce it to 10.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    makerCheckerHelper.actionAttachment(url, userAgent.getId());
+  }
+
+  @Then("$agent can verify the attachment is $state")
+  public void verifyStateOfMakerChecker(String agent, String state)
+      throws MessagesException, HealthCheckFailedException {
+    SymUser userAgent = userHelper.getUser(agent.toUpperCase());
+    Optional<SymMessage> message = messageHelper.getLatestTicketMessage(userAgent, initialTime);
+
+    assertEquals(state.toUpperCase(), getValueFromParam(message.get().getEntityData(), STATE));
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Please validate the approved/denied message posted in the ticket room.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/helpers/MakerCheckerHelper.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+   * @return ResponseEntity
+   */
+  public ResponseEntity actionAttachment(String actionUrl, Long agentId) throws MalformedURLException {
+    URL context = new URL(actionUrl);
+    String url = context.getPath().replace(HELPDESK_BOT, "");
+
+    UriComponentsBuilder builder = UriComponentsBuilder.fromUriString(url)
+        .queryParam(USER_ID, agentId);
+
+    ResponseEntity<TicketResponse> responseEntity =
```
Comment (robson-sanchez) [inline helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/helpers/MakerCheckerHelper.java]: It should be 'MakerCheckerResponse' class.

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 2 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 3 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 4 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 5 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/listener/PrepareEnvironmentListener.java (hunk 7 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/listener/PrepareEnvironmentListener.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/helpers/MakerCheckerHelper.java (hunk 9 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/helpers/MakerCheckerHelper.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 1 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 3 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 6 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 8 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

File: helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java (hunk 6 above)
Comment (crepache) [reply helpdesk-bot/src/test/java/org/symphonyoss/symphony/bots/helpdesk/bot/it/steps/MakerCheckerSteps.java]: Done.

**Review by robson-sanchez — APPROVED** [review]

**Pull request merged**